base/range.jl

```diff
@@ -296,19 +296,26 @@ unitrange_last(start::T, stop::T) where {T} =
                           convert(T,start-oneunit(stop-start)))
 
 if isdefined(Main, :Base)
+    function checkbounds(@nospecialize(t::Tuple), r::AbstractUnitRange)
+        isempty(r) && return
+        if !(1 <= first(r) <= length(t) && 0 <= last(r) <= length(t))
+            throw_boundserror(t, r)
+        end
+    end
     # Constant-fold-able indexing into tuples to functionally expose Base.tail and Base.front
-    function getindex(@nospecialize(t::Tuple), r::UnitRange)
+    function getindex(@nospecialize(t::Tuple), r::AbstractUnitRange)
         @_inline_meta
-        r.start > r.stop && return ()
-        if r.start == 1
-            r.stop == length(t)   && return t
-            r.stop == length(t)-1 && return front(t)
-            r.stop == length(t)-2 && return front(front(t))
-        elseif r.stop == length(t)
-            r.start == 2 && return tail(t)
-            r.start == 3 && return tail(tail(t))
+        isempty(r) && return ()
+        if first(r) == 1
+            last(r) == length(t)   && return t
+            last(r) == length(t)-1 && return front(t)
+            last(r) == length(t)-2 && return front(front(t))
+        elseif last(r) == length(t)
+            first(r) == 2 && return tail(t)
+            first(r) == 3 && return tail(tail(t))
         end
-        return (eltype(t)[t[ri] for ri in r]...,)
```

Comment:
I recall we did it this way originally for compiler performance, but maybe it's ok now?

Reply:
I didn't notice any big latencies at a quick glance, but that probably isn't saying much. I don't really expect specializing on tuples up to length 10 to be problematic for the compiler and for lengths greater than that, it should use this as a fallback minus the `eltype` annotations. Is it worth worrying about an extra layer of function call in there? Or is there a specific edge case you are worried about?

Reply:
You were right, I did measure some compile time overhead with the previous approach for very large tuples. Should be fixed now though.